src/WebJobs.Script/Diagnostics/FileTraceWriter.cs

```diff
@@ -39,90 +40,141 @@ public FileTraceWriter(string logFilePath, TraceLevel level, LogType logType) :
+
+        private protected virtual void StartFlushTimer()
         {
-            if (_logBuffer.Count == 0)
+            if (_flushTimer != null)
             {
                 return;
             }
 
-            ConcurrentQueue<string> currentBuffer = null;
-            lock (_syncLock)
+            lock (_flushTimerSyncRoot)
             {
-                // Snapshot the current set of buffered logs
-                // and set a new queue. This ensures that any new
-                // logs are written to the new buffer.
-                // We do this snapshot in a lock since Flush might be
-                // called by multiple threads concurrently, and we need
-                // to ensure we only log each log once.
-                currentBuffer = _logBuffer;
-                _logBuffer = new ConcurrentQueue<string>();
+                if (_flushTimer == null)
+                {
+                    // start a timer to flush accumulated logs in batches
+                    _flushTimer = new Timer(FlushCallback, null, LogFlushIntervalMs, LogFlushIntervalMs);
+                }
             }
+        }
 
-            if (currentBuffer.Count == 0)
+        private protected virtual void StopFlushTimer()
+        {
+            if (_flushTimer == null)
             {
                 return;
             }
 
-            // concatenate all lines into one string
-            StringBuilder sb = new StringBuilder();
-            string line = null;
-            while (currentBuffer.TryDequeue(out line))
+            lock (_flushTimerSyncRoot)
             {
-                sb.AppendLine(line);
+                if (_flushTimer != null)
+                {
+                    _flushTimer.Dispose();
+                    _flushTimer = null;
+                }
             }
+        }
 
-            if (_currentLogFileInfo == null)
-            {
-                // delay create log file
-                SetLogFile();
-            }
+        private void FlushCallback(object state)
+        {
+            Flush(true);
+        }
 
-            // write all lines in a single file operation
-            string content = sb.ToString();
-            try
-            {
-                AppendToFile(_currentLogFileInfo, content);
-            }
-            catch (Exception ex) when (ex is DirectoryNotFoundException || ex is FileNotFoundException)
+        public override void Flush()
+        {
+            Flush(false);
+        }
+
+        private protected virtual bool Flush(bool fromTimer)
+        {
+            if (_logBuffer.Count == 0)
             {
-                // need to handle cases where log files or directories might
-                // have been deleted from underneath us
-                SetLogFile();
-                AppendToFile(_currentLogFileInfo, content);
+                if (fromTimer)
+                {
+                    StopFlushTimer();
```

Comment:
As an optimization, we'll stop the timer if there is no work to do when a timer is fired. This is done here, rather than at the end, to avoid the expense of re-creating the timer more frequently.